### PR TITLE
#Fix #4738 - Workflow: Unify RunOptions/ForwardTranslator Options accross entire code base

### LINKED ISSUE #4738: Workflow: Unify RunOptions/ForwardTranslator Options accross entire code base
Enhancement Request
--------------

- filetypes/WorkflowJSON.hpp uses filetypes/RunOptions
- filetypes/RunOptions stores the ForwardTranslatorOptions as a json string which is highly inconvenient
- The CLI has flags to set these RunOptions/ForwardTranslatorOptions
- The E+ ForwardTranslator has individual methods to see what corresponds to the ForwardTranslator options

## Possible Implementation

Add a filetypes/ForwardTranslatorOptions class. Make RunOptions use that type directly.

Use the same classes accross the entire code base (in E+ ForwardTranslator and CLI)

The RunOptions/ForwardTranslatorOptions passed in the CLI should be **MERGED** with the WorkflowJSON ones (eg: anything that is NOT defaulted in the CLI should override the WorkflowJSON one (=user supplied values trump everything))

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------

- Fixes #4738

### Pull Request Author

 - [ ] Model API Changes / Additions
 - [ ] Any new or modified fields have been implemented in the EnergyPlus ForwardTranslator (and ReverseTranslator as appropriate)
 - [ ] Model API methods are tested (in `src/model/test`)
 - [ ] EnergyPlus ForwardTranslator Tests (in `src/energyplus/Test`)
 - [ ] If a new object or method, added a test in NREL/OpenStudio-resources: Add Link
 - [ ] If needed, added VersionTranslation rules for the objects (`src/osversion/VersionTranslator.cpp`)
 - [ ] Verified that C# bindings built fine on Windows, partial classes used as needed, etc.
 - [ ] All new and existing tests passes
 - [ ] If methods have been deprecated, update rest of code to use the new methods

**Labels:**

 - [ ] If change to an IDD file, add the label `IDDChange`
 - [ ] If breaking existing API, add the label `APIChange`
 - [ ] If deemed ready, add label `Pull Request - Ready for CI` so that CI builds your PR

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] Code Style, strip trailing whitespace, etc.
 - [ ] All related changes have been implemented: model changes, model tests, FT changes, FT tests, VersionTranslation, OS App
 - [ ] Labeling is ok
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified

### CODE REVIEW

**Comment by jmarrec** [comment]
I rebased the branch onto develop, since it was getting a bit stale but I want a clean commit history.

File: src/utilities/filetypes/RunOptions.hpp [hunk 1]
```diff
@@ -100,9 +110,13 @@ class UTILITIES_API RunOptions
...
-  std::string forwardTranslateOptions() const;
-  bool setForwardTranslateOptions(const std::string& options);
-  void resetForwardTranslateOptions();
+  OS_DEPRECATED std::string forwardTranslateOptions() const;
+  OS_DEPRECATED bool setForwardTranslateOptions(const std::string& options);
+  OS_DEPRECATED void resetForwardTranslateOptions();
+
+  ForwardTranslatorOptions forwardTranslatorOptions() const;
+  bool setForwardTranslatorOptions(const ForwardTranslatorOptions& forwardTranslatorOptions);
+  void resetForwardTranslatorOptions();
```
Comment (jmarrec) [inline src/utilities/filetypes/RunOptions.hpp:119]: Deprecated and new. Before it was storing everything as a json encoded string, now it's an explicit class.

Deprecated are kept for compat with ruby workflow-gem until it gets phased out

File: src/utilities/filetypes/RunOptions_Impl.hpp [hunk 2]
```diff
@@ -96,9 +86,9 @@ namespace detail {
     bool setCustomOutputAdapter(const CustomOutputAdapter& adapter);
     void resetCustomOutputAdapter();
 
-    std::string forwardTranslateOptions() const;
-    bool setForwardTranslateOptions(const std::string& options);
-    void resetForwardTranslateOptions();
+    ForwardTranslatorOptions forwardTranslatorOptions() const;
+    bool setForwardTranslatorOptions(const ForwardTranslatorOptions& options);
+    void resetForwardTranslatorOptions();
```
Comment (jmarrec) [inline src/utilities/filetypes/RunOptions_Impl.hpp:91]: Impl only keeps the new stuff

File: src/utilities/filetypes/test/WorkflowJSON_GTest.cpp [hunk 3]
```diff
@@ -1110,21 +1112,42 @@ TEST(Filetypes, RunOptions_ForwardTranslate) {
...
+  {
+    ForwardTranslatorOptions ftOptions = options.forwardTranslatorOptions();
+    EXPECT_FALSE(ftOptions.keepRunControlSpecialDays());
+    EXPECT_TRUE(ftOptions.iPTabularOutput());
+    EXPECT_TRUE(ftOptions.excludeLCCObjects());
+    EXPECT_TRUE(ftOptions.excludeSQliteOutputReport());
+    EXPECT_TRUE(ftOptions.excludeHTMLOutputReport());
+    EXPECT_TRUE(ftOptions.excludeVariableDictionary());
+    EXPECT_TRUE(ftOptions.excludeSpaceTranslation());
+  }
```
Comment (jmarrec) [inline src/utilities/filetypes/test/WorkflowJSON_GTest.cpp:1149]: backward compat test

File: src/utilities/filetypes/ForwardTranslatorOptions.hpp [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+  void setExcludeVariableDictionary(bool excludeVariableDictionary);
+  void resetExcludeVariableDictionary();
+
+  bool excludeSpaceTranslation() const;
+  bool isExcludeSpaceTranslationDefaulted() const;
+  void setExcludeSpaceTranslation(bool excludeSpaceTranslation);
+  void resetExcludeSpaceTranslation();
+
+  /* Any non-defaulted value from other is brought over */
+  void overrideValuesWith(const ForwardTranslatorOptions& other);
```
Comment (jmarrec) [inline src/utilities/filetypes/ForwardTranslatorOptions.hpp:104]: :eyes:

File: src/utilities/filetypes/ForwardTranslatorOptions.hpp [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+  // TODO: keep?
+  static std::vector<ForwardTranslatorOptionKeyMethod> forwardTranslatorOptionKeyMethods();
+
+ protected:
+  Json::Value json() const;
+  static ForwardTranslatorOptions fromJSON(const Json::Value& value);
```
Comment (jmarrec) [inline src/utilities/filetypes/ForwardTranslatorOptions.hpp:111]: to/from Jsoncpp

File: src/utilities/filetypes/ForwardTranslatorOptions_Impl.hpp [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+    bool m_is_no_sqlite_output_defaulted = true;
+
+    bool m_no_html_output = DEFAULT_NO_HTML_OUTPUT;
+    bool m_is_no_html_output_defaulted = true;
+
+    bool m_no_variable_dictionary = DEFAULT_NO_VARIABLE_DICTIONARY;
+    bool m_is_no_variable_dictionary_defaulted = true;
+
+    bool m_no_space_translation = DEFAULT_NO_SPACE_TRANSLATION;
+    bool m_is_no_space_translation_defaulted = true;
```
Comment (jmarrec) [inline src/utilities/filetypes/ForwardTranslatorOptions_Impl.hpp:126]: Single place to define the defaults + Handle the fact that stuff can be defaulted for overrideValuesWith

File: src/utilities/filetypes/ForwardTranslatorOptions.cpp [hunk 7]
```diff
@@ -0,0 +1,463 @@
...
+
+    m_no_html_output = DEFAULT_NO_HTML_OUTPUT;
+    m_is_no_html_output_defaulted = true;
+
+    m_no_variable_dictionary = DEFAULT_NO_VARIABLE_DICTIONARY;
+    m_is_no_variable_dictionary_defaulted = true;
+
+    m_no_space_translation = DEFAULT_NO_SPACE_TRANSLATION;
+    m_is_no_space_translation_defaulted = true;
+  }
```
Comment (jmarrec) [inline src/utilities/filetypes/ForwardTranslatorOptions.cpp:62]: provide a reset function

File: src/utilities/filetypes/ForwardTranslatorOptions.cpp [hunk 8]
```diff
@@ -0,0 +1,463 @@
...
+
+  void ForwardTranslatorOptions_Impl::setExcludeVariableDictionary(bool excludeVariableDictionary) {
+    m_no_variable_dictionary = excludeVariableDictionary;
+    m_is_no_variable_dictionary_defaulted = false;
+  }
+
+  void ForwardTranslatorOptions_Impl::resetExcludeVariableDictionary() {
+    m_no_variable_dictionary = DEFAULT_NO_VARIABLE_DICTIONARY;
+    m_is_no_variable_dictionary_defaulted = true;
+  }
```
Comment (jmarrec) [inline src/utilities/filetypes/ForwardTranslatorOptions.cpp:170]: getter, setters, isXXDefaulted and resetXXX are provided, similar to model objects. Except it's not IDD based but uses explicit members

File: src/utilities/filetypes/ForwardTranslatorOptions.cpp [hunk 9]
```diff
@@ -0,0 +1,463 @@
...
+    }
+
+    if (!other.isExcludeVariableDictionaryDefaulted()) {
+      setExcludeVariableDictionary(other.excludeVariableDictionary());
+    }
+
+    if (!other.isExcludeSpaceTranslationDefaulted()) {
+      setExcludeSpaceTranslation(other.excludeSpaceTranslation());
+    }
+  }
```
Comment (jmarrec) [inline src/utilities/filetypes/ForwardTranslatorOptions.cpp:218]: Merging stuff!

File: src/utilities/filetypes/ForwardTranslatorOptions.cpp [hunk 10]
```diff
@@ -0,0 +1,463 @@
...
+    // Write to string
+    Json::StreamWriterBuilder wbuilder;
+    // mimic the old StyledWriter behavior:
+    wbuilder["indentation"] = "   ";
+    std::string result = Json::writeString(wbuilder, json());
+
+    return result;
+  }
+
+}  // namespace detail
```
Comment (jmarrec) [inline src/utilities/filetypes/ForwardTranslatorOptions.cpp:265]: json and string serialization

File: src/utilities/filetypes/RunOptions.cpp [hunk 11]
```diff
@@ -330,28 +315,12 @@ boost::optional<RunOptions> RunOptions::fromString(const std::string& s) {
...
-    }
-
-    Json::StreamWriterBuilder wbuilder;
-    // mimic the old StyledWriter behavior:
-    wbuilder["indentation"] = "   ";
-    std::string optionString = Json::writeString(wbuilder, cleaned_options);
-
-    result->setForwardTranslateOptions(optionString);
+  if (value.isMember("ft_options") && value["ft_options"].isObject()) {
+    result.setForwardTranslatorOptions(ForwardTranslatorOptions::fromJSON(value["ft_options"]));
```
Comment (jmarrec) [inline src/utilities/filetypes/RunOptions.cpp:323]: Here's why the fromJSON static factory is needed

File: src/utilities/filetypes/RunOptions.cpp [hunk 12]
```diff
@@ -456,18 +425,44 @@ void RunOptions::resetCustomOutputAdapter() {
...
 }
 
 void RunOptions::resetForwardTranslateOptions() {
-  getImpl<detail::RunOptions_Impl>()->resetForwardTranslateOptions();
+  LOG(Warn, "As of 3.5.1, resetForwardTranslateOptions is deprecated. Use resetForwardTranslateOptions instead. "
+            "It will be removed within three releases.");
+  getImpl<detail::RunOptions_Impl>()->resetForwardTranslatorOptions();
 }
 
+/// @endcond
```
Comment (jmarrec) [inline src/utilities/filetypes/RunOptions.cpp:464]: Deprecated stuff.

File: src/energyplus/ForwardTranslator.hpp [hunk 13]
```diff
@@ -545,6 +538,13 @@ namespace energyplus {
    */
     std::vector<LogMessage> errors() const;
 
+    ForwardTranslatorOptions forwardTranslatorOptions() const;
+
+    void setForwardTranslatorOptions(ForwardTranslatorOptions forwardTranslatorOptions);
```
Comment (jmarrec) [inline src/energyplus/ForwardTranslator.hpp:543]: E+ FT uses the same class now. One less duplication.

File: src/energyplus/ForwardTranslator/ForwardTranslateEnergyManagementSystemActuator.cpp [hunk 14]
```diff
@@ -134,7 +134,7 @@ namespace energyplus {
             }
           }
           //Give WArning that spaceType has multiple spaces
-          if (m_excludeSpaceTranslation && spaceType->spaces().size() > 1) {
+          if (m_forwardTranslatorOptions.excludeSpaceTranslation() && spaceType->spaces().size() > 1) {
```
Comment (jmarrec) [inline src/energyplus/ForwardTranslator/ForwardTranslateEnergyManagementSystemActuator.cpp:137]: Update the usage of the FT options throughout the indiv ft functions

File: src/workflow/WorkflowRunOptions.hpp [hunk 15]
```diff
@@ -31,41 +31,41 @@
...
+  /*
+    bool m_debug = false;
+    bool m_epjson = false;
+    bool m_fast = false;
+    bool m_preserveRunDir = false;
+    bool m_skipExpandObjects = false;
+    bool m_skipEnergyPlusPreprocess = false;
+    ForwardTranslatorOptions m_forwardTranslatorOptions;
+    boost::optional<CustomOutputAdapter> m_customOutputAdapter
+*/
```
Comment (jmarrec) [inline src/workflow/WorkflowRunOptions.hpp:68]: The OSWorkflow uses a class WorkflowRunOptions with has:
* Workflow Specific things like I/O, etc
* The utilities/RunOptions class as a member. That eliminates the last duplication

File: src/workflow/OSWorkflow.cpp [hunk 16]
```diff
@@ -34,16 +35,41 @@ OSWorkflow::OSWorkflow(const filesystem::path& oswPath, ScriptEngineInstance& ru
...
+    auto ori_ftOptions = runOpt_->forwardTranslatorOptions();
+    workflowJSON.setRunOptions(t_workflowRunOptions.runOptions);
+    // user supplied CLI flags trump everything
+    ori_ftOptions.overrideValuesWith(t_workflowRunOptions.runOptions.forwardTranslatorOptions());
+    workflowJSON.runOptions()->setForwardTranslatorOptions(ori_ftOptions);
+  }
+
+  if (workflowJSON.runOptions()->debug()) {
+    fmt::print("workflowJSON={}\n", workflowJSON.string());
+  }
```
Comment (jmarrec) [inline src/workflow/OSWorkflow.cpp:72]: Merge CLI provided run & ft options with the WorkflowJSON (workflow.osw) use for the simulation. CLI Flags take priority over the WorkflowJSON ones.

**Comment by jmarrec** [comment]
Ha geez, windows test errors... I will have to build locally and see what's going on.

**Comment by jmarrec** [comment]
I think I have a race condition because I simulate using the regular CLI and the Labs one, so that uses the same output directory...

**Comment by jmarrec** [comment]
Not sure what's going on. I dowloaded the windows package: http://openstudio-ci-builds.s3-website-us-west-2.amazonaws.com/PR-4739/OpenStudio-3.5.1-alpha+8634e94006-Windows.tar.gz

This works fine:

```shell
.\bin\openstudio.exe labs run -w .\Examples\compact_osw\compact_python_only.osw
```

**Review by tijcolem — APPROVED** [review]

**Pull request merged**